proposals/NNNN-unlock-existential-types-for-all-protocols.md

```diff
@@ -0,0 +1,321 @@
+
+```swift
+protocol Foo {
+  associatedtype Bar
+
+  func foo(_: Bar) -> Bar
+}
+
+private extension Foo {
+  // Forward to the foo method in an existential-accessible way, asserting that
+  // the '_Bar' generic argument matches the actual 'Bar' associated type of the
+  // dynamic value.
+  func _fooThunk<_Bar>(_ bar: _Bar) -> _Bar {
+    assert(_Bar.self == Bar.self)
+    let result = foo(unsafeBitCast(bar, to: Bar.self))
+    return unsafeBitCast(result, to: _Bar.self)
+  }
+}
+
+struct AnyFoo<Bar>: Foo {
+  private var _value: Foo
+
+  init<F: Foo>(_ value: F) where F.Bar == Bar {
+    self._value = value
+  }
+  
+  func foo(_ bar: Bar) -> Bar {
+    return self._value._fooThunk(bar)
+  }
+}
+```
+
+## Proposed Solution
+
+We suggest allowing any protocol to be used as a type and exercise the restriction on individual member accesses uniformly across extension members and requirements. Additionally, the adjusted access criteria for protocol members shall account for associated types with known implementations:
+
+```swift
+protocol IntCollection: RangeReplaceableCollection where Self.Element == Int {}
+
+let array: IntCollection = [3, 1, 4, 1, 5]
+
+array.append(9) // OK, 'Self.Element' is known to be 'Int'.
+```
+
+Having lowered the limitation, the mere presence of an associated type requirement will no longer preclude member accesses, but references to `Self`-rooted associated types *will* for the same reasons some `Self` references do today. As alluded to back in [Inconsistent Language Semantics](#inconsistent-language-semantics), references to covariant `Self` are already getting automatically replaced with the base object type, permitting usage of `Self`-returning methods on existential values:
+
+```swift
+protocol Copyable {
+  func copy() -> Self
+}
+
+func test(_ c: Copyable) {
+  let x = c.copy() // OK, x is of type 'Copyable'
+}
+```
+
+Because they tend to outnumber direct uses of `Self` in protocol contexts, and for the sake of consistency, we believe that extending covariant type erasure to associated types is a reasonable undertaking in light of the primary focus:
+
+```swift
+func test(_ collection: RandomAccessCollection) {
+  // func dropLast(_ k: Int = 1) -> SubSequence
+  let x = collection.dropLast() // OK, x is of type 'RandomAccessCollection'
+}
+```
+___
+
+This way, a protocol or protocol extension member (method/property/subscript/initializer) may be used on an existential value iff the following criteria hold:
+* The type of the invoked member (accessor — for storage declarations), as viewed in context of the *base type*, must **not** contain references to `Self` or `Self`-rooted associated types in [non-covariant](https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)) position.
+
+> The "`isAvailableOnExistential`" routine will consider the following types covariant:
```

Comment:
Could we replace the first bit with something like `The compiler will consider` instead of `The isAvailableOnExistential routine will consider`? Given we're talking about implementation details here and the name of the routine can always change in the future, so it would be better to be more general here.

Reply:
The name was meant to refer to some piece of code and its purpose, hence the quotes. I chose to be specific here because the constraint system (a different compiler component) and its covariant coercions are a superset of this list. How about we switch to indefinite passive and say "the following types will be considered covariant"?

Reply:
That sounds good to me!